### PR TITLE
feat: hello-world namespace support for ADVZ VID scheme

### PR DESCRIPTION
**Pull request by ggutoski**
closes: #388 

# Revised 2023-11-06

Force-pushed update incorporating comments from previous reviews and offline discussion.

- Add `PayloadProver` subtrait for `VidScheme`. It's a generic trait with two implementations for ADVZ. See rustdoc: https://github.com/EspressoSystems/jellyfish/blob/93e396ebed1421458e2647d6461932dffb44680d/primitives/src/vid/advz/payload_prover.rs#L9-L15
- Change `VidScheme` to use `&[u8]` for payload type as per offline discussion.
- move `bytes_to_field` from `conversion.rs` to a new `bytes_to_field.rs` file owned by `advz`.

# Excerpts from original description

- change the bytes-to-field conversion in `conversion.rs`. It's much simpler now because I decided we don't need to add metadata into the output field elements. Instead, it's the responsibility of the caller to know where exactly their bytes begin/end in the encoded field elements.
- `advz.rs` has a large diff but it's mostly just moving code around.
- fix a bug in encoding payload bytes to polynomials wherein FFT and inverse-FFT were accidentally swapped.
- fix some nix stuff 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: primitives/src/vid/namespace.rs
Comment (jbearer) [inline primitives/src/vid/namespace.rs:35]: Are `commit` and `common` always together? Would it make sense to just put all the common stuff in `commit`?

File: primitives/src/vid/namespace.rs
Comment (jbearer) [inline primitives/src/vid/namespace.rs:24]: The prover (e.g. query server) also needs some way to get the actual data. So I think this should be something like `VidResult<(Vec<u8>, Self::DataProof)>`, or `Self::DataProof` should have some trait bound that allows me to extract the data from the proof

File: primitives/src/vid/namespace.rs
Comment (jbearer) [inline primitives/src/vid/namespace.rs:46]: I thought we discussed not having any namespace-specific interface in Jellyfish, but having the sequencer itself use the namespace table to convert from namespace index to byte range, and then just use the above APIs to read bytes?

File: primitives/src/vid/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/namespace.rs:35]: `commit` is supposed to be an extremely small item (like a 32-byte hash).

`common` is metadata sent to all storage nodes such as KZG commitments to all the polynomials in a block payload. It's size is not bounded. I view it more as part of a proof to be verified against `commit` than as part of `commit` itself.

File: primitives/src/vid/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/namespace.rs:24]: My understanding is that the query server will already have `payload` in full. Thus, it already has the actual data---it could grab its own subslice of `payload`.

But this raises a related issue: Due to bytes-to-field encoding, the user needs a few bytes outside the tx/namespace in order to verify the proof. Where should these bytes go? I guess it makes sense to put those few bytes inside the `DataProof`.

But it's not clear to me that the actual data should also go into the `DataProof`.

File: primitives/src/vid/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/namespace.rs:46]: My takeaway from that discussion is the rule "no namespace-specific stuff in jellyfish, instead just give me a byte range" applies only to  *transactions*.

Things are a bit different for a full namespace proof. I do suspect that `namespace_verify` in jellyfish will never be used except in tests for correctness. So basically this function is a reference implementation against which to test smart contract verification of a namespace. wdyt?

File: primitives/src/vid/advz.rs
Comment (chancharles92) [inline primitives/src/vid/advz.rs:210]: `coeffs_iter` -> `evals_iter`? (the elements are no longer coefficients.)

File: primitives/src/vid/advz.rs
Comment (chancharles92) [inline primitives/src/vid/advz.rs:227]: Same here, maybe better to change the variable name of `coeffs_iter` to `evals_iter`.

File: primitives/src/vid/advz.rs
Comment (chancharles92) [inline primitives/src/vid/advz.rs:285]: Can we derive `pseudorandom_scalar` from `commit` + `all_evals_digest`? There can be redundant hash calls if we derive from `poly_commits`.

File: primitives/src/vid/advz/namespace.rs
Comment (chancharles92) [inline primitives/src/vid/advz/namespace.rs:69]: Optional TODO in the future PR: save this computation.

File: primitives/src/vid/advz/namespace.rs
Comment (chancharles92) [inline primitives/src/vid/advz/namespace.rs:39]: In the namespace application, does the prover need to generate these proofs online? Can it just batchly preprocess all of the evaluation proofs offline, and in the online phase, whenever a client asks, it just forwards proofs without doing any computations?
cc @ggutoski @jbearer

Comment (ggutoski) [reply primitives/src/vid/advz/namespace.rs:39]: Yes, it could. I suspect servers will opt to pre-process everything if they get sufficient traffic. Seems to me that the the priority is to optimize data on the wire, so KZG batch proofs should be quite helpful here.

File: primitives/src/vid/namespace.rs
Comment (jbearer) [reply primitives/src/vid/namespace.rs:35]: Gotcha

File: primitives/src/vid/namespace.rs
Comment (jbearer) [reply primitives/src/vid/namespace.rs:24]: Yeah, putting the extra bytes in `DataProof` and having the payload data separate seems fine. But the query server still needs some interface for actually reading bytes from the full payload. I don't see that here since `Self::Payload` doesn't have a trait bound that would allow me to read bytes out of it

Comment (ggutoski) [reply primitives/src/vid/namespace.rs:24]: Good point. I'll add something for that.

Comment (ggutoski) [reply primitives/src/vid/namespace.rs:24]: As discussed offline I'll remove the `Payload` assoc type and just enforce `&[u8]` payload. It hurts my soul to be so opinionated in the API, but we discussed that downstream users are assumed to bear the cost of serializing to a 1D byte slice, and this assumption makes our life much easier.

File: primitives/src/vid/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/namespace.rs:46]: As discussed offline, both namespace and tx proofs will have the exact same API. I'll make a generic trait and implement it twice: once for namespace proofs (snark-friendly) and once for tx proofs (requires a pairing). There will be no namespace-specific stuff in jellyfish. Instead the API is just "give me a index range and I'll give you a proof."

File: primitives/src/vid/advz.rs
Comment (ggutoski) [reply primitives/src/vid/advz.rs:210]: changed `coeffs_iter` -> `evals_iter` because it's (an iterator over) the evaluations of a polynomial.

File: primitives/src/vid/advz.rs
Comment (ggutoski) [reply primitives/src/vid/advz.rs:285]: Yep. Done in commit 8f2240b3f9d01c29f77142c8b2dd8698c0be79fe

File: primitives/src/vid/advz/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/advz/namespace.rs:69]: #397

File: primitives/src/vid/advz/namespace.rs
Comment (ggutoski) [reply primitives/src/vid/advz/namespace.rs:39]: #397

File: primitives/src/vid/payload_prover.rs
Comment (jbearer) [inline primitives/src/vid/payload_prover.rs:25]: I think `range` should also be a public input here. The statement being verified is not just `chunk` is a slice from `commit`, but `chunk` is _the_ slice at position `range` in `commit`.

File: primitives/src/vid/advz/payload_prover.rs
Comment (jbearer) [inline primitives/src/vid/advz/payload_prover.rs:9]: It would be nice to give type aliases for these different proof types in terms of the high level properties/use cases, that make the contrasts obvious. I don't have any great ideas, though. Maybe something like `GranularPayloadProof`/`RangePayloadProof` or something?

File: primitives/src/vid/advz/payload_prover.rs
Comment (jbearer) [inline primitives/src/vid/advz/payload_prover.rs:30]: Something slightly annoying about this interface is as a downstream user I'm probably going to have to write out whatever `P::Proof` is in order to help the typechecker figure out which kind of proof I want, since there are two different implementations with the same interface. In practice I'm probably going to be referring to the VID scheme itself through a type alias with specific instantiations of `<P, T, H, V>`. Makes it kind of annoying to click through rust docs to try and figure out what `P` is and then what `P::Proof` is so I can name this `Proof<P::Proof>` type.

Two ideas for making this a bit more ergonomic:

1. Bring back the monolothic prover trait with two pairs of methods with identical interfaces, but different names. That way it's obvious from the name of the method which type of proof I'm asking for, I don't have to actually write down the type. However, this would enshrine in the trait interface the idea that we have these two separate types of proofs, which might not be true for every conceivable VID scheme.
2. Introduce a marker type which we use to denote the kind of proof we're looking for abstractly, decoupled from the specific type we use to implement that proof. The marker type can be used to select the trait implementation I want and the actual type of proof can be an associated type of that trait. Something roughly like this:

```rust
trait PayloadProver<ProofType>: VidScheme {
  type Proof;
  fn payload_proof(...) -> VidResult<Self::Proof>;
}

struct SnarkFriendlyProof;
struct GranularProof;

impl PayloadProver<GranularProof> for GenericAdvz<P, T, H, V> {
  type Proof = Proof<P::Proof>;
  ...
}

impl PayloadProver<GranularProof> for GenericAdvz<P, T, H, V> {
  type Proof = CommitRecovery<P::Evaluation>;
  ...
}
```

Use like
```rust
PayloadProver::<GranularProof>::payload_proof(&scheme, payload, range);
PayloadProver::<SnarkFriendlyProof>::payload_proof(&scheme, payload, range);
```

File: primitives/src/pcs/univariate_kzg/mod.rs
Comment (chancharles92) [inline primitives/src/pcs/univariate_kzg/mod.rs:95]: Minor: we can uncomment it but add a flag-feature so that it doesn't execute when we do disperse benchmark.

File: primitives/src/vid/payload_prover.rs
Comment (ggutoski) [reply primitives/src/vid/payload_prover.rs:29]: It's assumed that the `range` data is included in `proof`. Both impls for this trait include a `chunk_range: Range<usize>,` field. It's sufficient to pass tests. I prefer it this way to reduce clutter in the args. Sound ok to you?

File: primitives/src/vid/advz.rs
Comment (chancharles92) [inline primitives/src/vid/advz.rs:311]: We can omit this timer.

File: primitives/src/vid/advz/payload_prover.rs
Comment (ggutoski) [reply primitives/src/vid/advz/payload_prover.rs:30]: I like the marker type idea and I think we should use it if this is indeed a problem. But are you sure it will be so bad for downstream users? My tests work with the following:
```rust
let data_proof: Proof<_> = advz.payload_proof(payload, range)?;
let chunk_proof: CommitRecovery<_> = advz.payload_proof(payload, range)?;
```
Admittedly the names `Proof` and `CommitRecovery` kinda suck. It's hard to come up with good names without mentioning "namespace" or "transaction". 😉

File: primitives/src/vid/advz/payload_prover.rs
Comment (chancharles92) [reply primitives/src/vid/advz/payload_prover.rs:9]: Agree, it's better to distinguish these two types of implementations/APIs.

File: primitives/src/vid/payload_prover.rs
Comment (jbearer) [reply primitives/src/vid/payload_prover.rs:29]: I think morally it should be an argument because it's part of the statement being proven, conceptually, not part of the proof. Otherwise as a verifier, I need some way to extract the `chunk_range` from the proof to check it against the expected range, which needs to be incorporated into a trait somehow to make sense with the rest of the trait-based interface. And, we introduce a footgun where the verifier forgets to do this extra check, which is much harder to do if the expected range is a required parameter.

File: primitives/src/vid/advz/payload_prover.rs
Comment (jbearer) [reply primitives/src/vid/advz/payload_prover.rs:30]: Ah the `<_>` is a nice trick. I think that's fine then

File: primitives/src/pcs/univariate_kzg/mod.rs
Comment (ggutoski) [reply primitives/src/pcs/univariate_kzg/mod.rs]: Oops! Forgot to uncomment this. Add `kzg-print-trace` feature flag in https://github.com/EspressoSystems/jellyfish/pull/389/commits/a0678336d0061f4f6fb2406401a5cc2179191c2e . I added it only to those timers that were causing me noise. I could add it everywhere else, too, but it does pollute the code.

File: primitives/src/vid/advz.rs
Comment (ggutoski) [reply primitives/src/vid/advz.rs]: done in https://github.com/EspressoSystems/jellyfish/pull/389/commits/a0678336d0061f4f6fb2406401a5cc2179191c2e

**Review by chancharles92 — COMMENTED** [review]
LGTM, can approve after addressing Jeb's concern about APIs.

File: primitives/src/vid/payload_prover.rs
Comment (ggutoski) [reply primitives/src/vid/payload_prover.rs]: New `Statement` struct in https://github.com/EspressoSystems/jellyfish/pull/389/commits/f399a4fbc7903b3b9d1c39df86ae7e248b6cd893

It's not needed in `payload_proof` but it does improve the readability of `payload_verify`.

I'll leave it to you to resolve this conversation when you're satisfied.

File: primitives/src/vid/advz/payload_prover.rs
Comment (ggutoski) [reply primitives/src/vid/advz/payload_prover.rs:9]: Renamed to `SmallRangeProof`, `LargeRangeProof` and tidied in https://github.com/EspressoSystems/jellyfish/pull/389/commits/eb8fde47d2e88a7dd27c2e8f88c9df6e75f2564d .

Unfortunately it's not clear how to give a useful type alias, as I would need to specify the type parameter `P`. But the whole file is generic over `P` so any type alias would need to hard-code a specific choice for `P`, which I don't want to do.

Comment (jbearer) [reply primitives/src/vid/advz/payload_prover.rs:9]: You could do `type SmallRangeProof<P> = Proof<<P as Trait>::Proof>`. But renaming is also good 👍

File: primitives/src/vid/advz/payload_prover.rs
Comment (jbearer) [inline primitives/src/vid/advz/payload_prover.rs:39]: The proof types should at least derive `Clone`, `Debug`, and serde traits

File: primitives/src/vid/payload_prover.rs
Comment (jbearer) [inline primitives/src/vid/payload_prover.rs:23]: Comment needs to be updated,  we no longer have `chunk`

**Review by jbearer — APPROVED** [review]
Looks good. I think we might be missing some important traits on the proof types, particularly serdes. But if you want we can get this merged and add those later. Or add them now. Either way

File: primitives/src/vid/advz/payload_prover.rs
Comment (ggutoski) [reply primitives/src/vid/advz/payload_prover.rs:39]: Definitely true. I'm going to merge now anyway because (i) this PR is already too big, (ii) don't want to initiate another round of approval, (iii) there will certainly be back-and-forth over issues like this as we integrate into the sequencer.

**Pull request merged**
